File: Problems/Course Schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        vector<int> visited(numCourses, 0); // 0 = unvisited, 1 = visiting, 2 = visited

        for (auto& pre : prerequisites) {
            graph[pre[1]].push_back(pre[0]);
        }

        function<bool(int)> dfs = [&](int course) {
            if (visited[course] == 1) return false; // cycle detected
            if (visited[course] == 2) return true;

            visited[course] = 1;
            for (int next : graph[course]) {
                if (!dfs(next)) return false;
            }
            visited[course] = 2;
            return true;
        };

        for (int i = 0; i < numCourses; ++i) {
            if (!dfs(i)) return false;
        }
        return true;
    }
};

File: Problems/Course Schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        int[] indeg = new int[numCourses];
        for (int i = 0; i < numCourses; i++) graph.add(new ArrayList<>());

        // build graph and indegrees
        for (int[] p : prerequisites) {
            int a = p[0], b = p[1];
            graph.get(b).add(a); // b -> a
            indeg[a]++;
        }

        // seed queue with zero indegree
        Queue<Integer> q = new ArrayDeque<>();
        for (int i = 0; i < numCourses; i++) if (indeg[i] == 0) q.add(i);

        int taken = 0;
        while (!q.isEmpty()) {
            int u = q.poll();
            taken++; // take this course
            for (int v : graph.get(u)) {
                if (--indeg[v] == 0) q.add(v); // unlock next course
            }
        }

        return taken == numCourses; // all courses taken means no cycle
    }
}
